src/lang/builtins_error.ml

```diff
@@ -22,49 +22,82 @@
 
 open Lang_builtins
 
-module Error = Lang.MkAbstract (struct
-  type content = string
+type error = { kind : string; msg : string option }
+
+module ErrorDef = struct
+  type content = error
 
   let name = "error"
-  let descr = Printf.sprintf "error(type=%s)"
+
+  let descr { kind; msg } =
+    Printf.sprintf "error(kind=%S,message=%s)" kind
+      (match msg with Some msg -> Printf.sprintf "%S" msg | None -> "none")
+
   let compare = Stdlib.compare
-end)
+end
+
+module Error = Lang.MkAbstract ((ErrorDef : Lang.AbstractDef))
 
 let () = Lang.add_module "error"
 
 let () =
   add_builtin "error.register" ~cat:Liq
-    ~descr:"Return an error of the given type"
-    [("", Lang.string_t, None, Some "Type of the error")] Error.t (fun p ->
-      let s = Lang.to_string (List.assoc "" p) in
-      Error.to_value s)
+    ~descr:"Register an error of the given Kind"
+    [("", Lang.string_t, None, Some "Kind of the error")] Error.t (fun p ->
+      let kind = Lang.to_string (List.assoc "" p) in
+      Error.to_value { kind; msg = None })
 
 let () =
-  add_builtin "error.type" ~cat:Liq ~descr:"Get the type of an error"
+  add_builtin "error.kind" ~cat:Liq ~descr:"Get the kind of an error"
     [("", Error.t, None, None)] Lang.string_t (fun p ->
-      Lang.string (Error.of_value (List.assoc "" p)))
+      Lang.string (Error.of_value (List.assoc "" p)).kind)
+
+let () =
+  add_builtin "error.message" ~cat:Liq ~descr:"Get the message of an error"
+    [("", Error.t, None, None)] (Lang.maybe_t Lang.string_t) (fun p ->
+      match Error.of_value (List.assoc "" p) with
+        | { msg = Some msg } -> Lang.string msg
+        | { msg = None } -> Lang.nothing)
 
 let () =
   add_builtin "error.raise" ~cat:Liq ~descr:"Raise an error."
-    [("", Lang.string_t, None, Some "Description of the error.")]
-    (Lang.univ_t ()) (fun p ->
-      let s = Lang.to_string (Lang.assoc "" 1 p) in
-      raise (Lang_values.Runtime_error ([], s)))
+    [
+      ("", Error.t, None, Some "Error kind.");
+      ( "",
+        Lang.maybe_t Lang.string_t,
+        Some Lang.nothing,
+        Some "Description of the error." );
+    ]
+    (Lang.univ_t ())
+    (fun p ->
+      let { kind } = Error.of_value (Lang.assoc "" 1 p) in
+      let msg =
+        Utils.maybe Lang.to_string (Lang.to_option (Lang.assoc "" 2 p))
+      in
+      raise (Lang_values.Runtime_error { Lang_values.kind; msg; pos = [] }))
 
 let () =
   let a = Lang.univ_t () in
-  add_builtin "error.catch" ~cat:Liq
+  add_builtin "error.catch" ~cat:Liq ~flags:[Lang.Hidden]
     ~descr:"Execute a function, catching eventual exceptions."
     [
-      ("", Lang.fun_t [] a, None, Some "Function to execute.");
-      ( "",
-        Lang.fun_t [(false, "", Lang.string_t)] a,
+      ( "errors",
+        Lang.list_t Error.t,
         None,
-        Some "Error handler." );
+        Some "Kinds of errors to catch. Catches all errors if empty." );
```

Comment:
```suggestion
        Some "Kinds of errors to catch. );
```
I am not so fond of non-monotonous behaviors like this. Someday someone will use this lis in a complicated way and will have bugs because the empty list is not the expected limit case. I would rather have a list option here and have `nothing` catching all.

Reply:
Just did it!